File: docs/static/examples/src/fireworks/java/Fireworks.java
import io.github.bernhardfritz.p8g.Sketch;
import org.joml.Vector2f;
import org.joml.Vector4f;

import java.util.HashSet;
import java.util.Set;

public class Fireworks extends Sketch {

    Set<Firework> fireworks = new HashSet<>();
    Vector2f gravity = new Vector2f(0.f, 0.2f);

    @Override
    public void draw() {
        background(new float[] { 0.f, 0.f, 0.f, 25.5f});
        if (random() < 0.1f) {
            Firework firework = new Firework(
              new Particle(
                      new Vector2f(random(width), height),
                      1.f,
                      new Vector2f(0.f, random(-9.f, -10.f)),
                      -1,
                      new Vector4f(255.f, 0.f, 0.f, 255.f)
              )
            );
            fireworks.add(firework);
        }
        for (Firework firework : fireworks) {
            firework.applyForce(gravity);
            firework.update();
            firework.draw();
        }
        fireworks.removeIf(Firework::isDone);
    }

    public static void main(String[] args) {
        Sketch sketch = new Fireworks();
        sketch.run(320, 320, "Fireworks");
    }
}

File: docs/static/examples/src/fireworks/js/fireworks.js
import p8g, { background, createCanvas, height, random, width } from 'p8g.js';
import Victor from 'victor';
import Firework from './firework';
import Particle from './particle';

let fireworks = [];
const gravity = new Victor(0, 0.2);

p8g.draw = () => {
  background([0, 0, 0, 25.5]);
  if (random() < 0.1) {
    const firework = new Firework(
      new Particle(
        new Victor(random(width), height),
        1,
        new Victor(0, random(-9, -10)),
        -1,
        [255, 0, 0, 255],
      ),
    );
    fireworks.push(firework);
  }
  for (const firework of fireworks) {
    firework.applyForce(gravity);
    firework.update();
    firework.draw();
  }
  let index = -1;
  while (
    (index = fireworks.findIndex((firework) => firework.isDone())) !== -1
  ) {
    fireworks.splice(index, 1);
  }
};

const canvasWrapper = document.getElementById('canvas-wrapper');
const canvas = createCanvas(320, 320);
canvasWrapper.appendChild(canvas);
